Mathlib/Analysis/Normed/Group/SeparationQuotient.lean

```diff
@@ -0,0 +1,231 @@
+/-
+Copyright (c) 2024 Yoh Tanimoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yoh Tanimoto
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.Normed.Group.Hom
+
+-- TODO modify doc, check if instances are really needed, golf
+-- want to define liftCLM. problem: a CLM on a normed vector space
+-- is not automatically a `NormedAddGroupHom
+
+
+/-!
+# Lifts of maps to separation quotients of seminormed groups
+
+For any `SeminormedAddCommGroup M`, a `NormedAddCommGroup` instance has been defined in
+`Mathlib.Analysis.Normed.Group.Uniform`.
+
+## Main definitions
+
+We use `M` and `N` to denote seminormed groups.
+All the following definitions are in the `NullSubgroup` namespace. Hence we can access
+`NullSubgroup.normedMk` as `normedMk`.
+
+* `normedAddCommGroupQuotient` : The normed group structure on the quotient by the null subgroup.
+    This is an instance so there is no need to explicitly use it.
+
+* `normedMk` : the normed group hom from `M` to `SeparationQuotient M`.
+
+* `lift f hf`: implements the universal property of `SeparationQuotient M`. Here
+    `(f : NormedAddGroupHom M N)`, `(hf : ∀ s ∈ nullSubgroup M, f s = 0)` and
+    `lift f hf : NormedAddGroupHom (SeparationQuotient M) N`.
+
+* `IsQuotient`: given `f : NormedAddGroupHom M N`, `IsQuotient f` means `N` is isomorphic
+    to a quotient of `M` by the null subgroup, with projection `f`. Technically it asserts `f` is
+    surjective and the norm of `f x` is the infimum of the norms of `x + m` for `m` in `f.ker`.
+
+## Main results
+
+* `norm_normedMk` : the operator norm of the projection is `1` if the subspace is not dense.
+
+* `IsQuotient.norm_lift`: Provided `f : normed_hom M N` satisfies `IsQuotient f`, for every
+     `n : N` and positive `ε`, there exists `m` such that `f m = n ∧ ‖m‖ < ‖n‖ + ε`.
+
+
+## Implementation details
+
+For any `SeminormedAddCommGroup M`, we define a norm on `SeparationQuotient M` by
+`‖x‖ = ‖mk x‖` using the lift.
+
+-/
+
+
+noncomputable section
+
+open SeparationQuotient Metric Set Topology NNReal
+
+variable {M N : Type*} [SeminormedAddCommGroup M] [SeminormedAddCommGroup N]
+
+namespace SeparationQuotientAddGroup
+
+/-- The null subgroup with respect to the norm. -/
+def nullSubgroup : AddSubgroup M where
+  carrier := {x : M | ‖x‖ = 0}
+  add_mem' {x y} (hx : ‖x‖ = 0) (hy : ‖y‖ = 0) := by
+    apply le_antisymm _ (norm_nonneg _)
+    refine (norm_add_le x y).trans_eq ?_
+    rw [hx, hy, add_zero]
+  zero_mem' := norm_zero
+  neg_mem' {x} (hx : ‖x‖ = 0) := by
+    simp only [mem_setOf_eq, norm_neg]
+    exact hx
+
+lemma inseparable_iff_norm_zero (x y : M) : Inseparable x y ↔ ‖x - y‖ = 0 := by
+  rw [Metric.inseparable_iff, dist_eq_norm]
+
+lemma isClosed_nullSubgroup : IsClosed (@nullSubgroup M _ : Set M) := by
```

Comment:
A much shorter proof is
```
  have : nullSubgroup (M := M) = closedBall (0 : M) 0 := by
    ext
    rw [mem_closedBall_zero_iff]
    sorry
  rw [this]
  exact isClosed_ball
```
you might find it easier to define the null subgroup as the place where `‖x‖ ≤ 0`, as then you can drop all the `le_antisymm`s in the proofs.

Reply:
A *much* shorter proof is:
```
  isClosed_singleton.preimage continuous_norm
```

Reply:
changed!